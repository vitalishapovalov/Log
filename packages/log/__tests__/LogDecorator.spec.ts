import {
    T,
    TConstructor,
    TGetPrototypeOf,
    TIsExtensible,
    TOwnKeys,
    TPreventExtensions,
    TSetPrototypeOf,
    TLogFullProperty,
    TFunction,
    TProvideLogger
} from "./__mocks__/LogDecorator";

// TODO when used as a parameter decorator
// TODO when used as a method decorator
// TODO when used as a getset decorator
// TODO when used on a class + property/method/getset
// TODO static properties/methods/getset
// TODO cover different options cases

describe("log/Log test", () => {

    describe("when used as a class decorator", () => {

        let dummy;
        let logSpy;

        beforeEach(() => {
           logSpy = jest.spyOn(console, "log");
        });

        afterEach(() => {
            logSpy.mockClear();
        });

        it("shouldn't log anything", () => {
            const inst = new T();

            dummy = Object.getPrototypeOf(inst);
            Object.isExtensible(inst);
            Object.setPrototypeOf(inst, Object.create({}));
            Object.preventExtensions(inst);

            expect(logSpy).not.toHaveBeenCalled();
        });

        it("should correctly log only get/set events on property", () => {
            const inst = new T();

            dummy = inst.prop;
            inst.prop = "";

            // shouldn't be logged
            dummy = "prop" in inst;
            delete inst.prop;
            Object.defineProperty(inst, "prop", { value: "prop_value" });
            dummy = Object.getOwnPropertyDescriptor(inst, "prop");

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(JSON.stringify(logSpy.mock.calls[1])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(2);
        });

        it("should correctly log only get/set events on getters and setters", () => {
            const inst = new T();

            dummy = inst.getSetProp;
            inst.getSetProp = "";

            // shouldn't be logged
            dummy = "getSetProp" in inst;
            delete inst.getSetProp;
            Object.defineProperty(inst, "getSetProp", { value: "getSetProp_value" });
            dummy = Object.getOwnPropertyDescriptor(inst, "getSetProp");

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(JSON.stringify(logSpy.mock.calls[1])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(2);
        });

        it("should correctly log only call events on methods", () => {
            const inst = new T();

            dummy = inst.method();

            // shouldn't be logged
            dummy = "method" in inst;
            delete inst.method;
            Object.defineProperty(inst, "method", { value: () => {} });
            dummy = Object.getOwnPropertyDescriptor(inst, "method");

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log constructor", () => {
            new TConstructor({ a: 10 }, 100, "100", ["str"]);

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log getPrototypeOf", () => {
            Object.getPrototypeOf(new TGetPrototypeOf());

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log setPrototypeOf", () => {
            Object.setPrototypeOf(new TSetPrototypeOf(), {
                dummyMethod() {}
            });

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log isExtensible", () => {
            Object.isExtensible(new TIsExtensible());

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log preventExtensions", () => {
            Object.preventExtensions(new TPreventExtensions());

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log ownKeys", () => {
            const instance = new TOwnKeys();

            Object.getOwnPropertyNames(instance);
            Object.getOwnPropertySymbols(instance);
            Reflect.ownKeys(instance);
            Object.keys(instance);
            for (const k in instance) {}

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(JSON.stringify(logSpy.mock.calls[1])).toMatchSnapshot();
            expect(JSON.stringify(logSpy.mock.calls[2])).toMatchSnapshot();
            expect(JSON.stringify(logSpy.mock.calls[3])).toMatchSnapshot();
            expect(JSON.stringify(logSpy.mock.calls[4])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(5);
        });

        it("should correctly log delete", () => {
            delete new TLogFullProperty().prop;

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log defineProperty", () => {
            Object.defineProperty(new TLogFullProperty(), "prop", { value: "new val" });

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log has", () => {
            dummy = "prop" in new TLogFullProperty();

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log getOwnPropertyDescriptor", () => {
            Object.getOwnPropertyDescriptor(new TLogFullProperty(), "prop");

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        it("should correctly log apply", () => {
            const instance = new TFunction("a", "b", "return a + b;");
            instance.call({}, 20, 30);

            expect(JSON.stringify(logSpy.mock.calls[0])).toMatchSnapshot();
            expect(logSpy).toHaveBeenCalledTimes(1);
        });

        describe("instance message logger", () => {

            it("should provide instance logger", () => {
               expect((new TProvideLogger() as any).logger).toBeTruthy();
            });

            it("should correctly log messages", () => {
                const logger = (new TProvideLogger() as any).logger;

                logger.log("");
                logger.info("Message []");
                logger.warn("Number: $number(1000)");
                logger.error("String: $string(alex)");

                expect(logSpy).toHaveBeenCalledTimes(4);
            });
        });
    });
});
