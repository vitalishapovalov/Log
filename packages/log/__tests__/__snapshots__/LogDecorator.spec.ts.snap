// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log/Log test when used as a class decorator should correctly log apply 1`] = `"[\\"\\\\n  Logger         @  TFunction                                       \\\\n  Call function  @  TFunction(arg1: Number, arg2: Number): Number;  \\\\n  Call context   @  {}                                              \\\\n  Call with      @  (20, 30)                                        \\\\n  Call result    @  50                                              \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log constructor 1`] = `"[\\"\\\\n  Logger                  @  TConstructor                      \\\\n  Construct new instance  @  Object                            \\\\n  Construct arguments     @  ({ a: 10 }, 100, \\\\\\"100\\\\\\", [\\\\\\"str\\\\\\"])  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log defineProperty 1`] = `"[\\"\\\\n  Logger                  @  TLogFullProperty      \\\\n  Call defineProperty on  @  TLogFullProperty      \\\\n  Property key            @  prop                  \\\\n  Property descriptor     @  { value: \\\\\\"new val\\\\\\" }  \\\\n  Call result             @  true                  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log delete 1`] = `"[\\"\\\\n  Logger                  @  TLogFullProperty        \\\\n  Access delete operator  @  delete Object[\\\\\\"prop\\\\\\"];  \\\\n  Delete result           @  true                    \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log getOwnPropertyDescriptor 1`] = `"[\\"\\\\n  Logger                            @  TLogFullProperty                              \\\\n  Call getOwnPropertyDescriptor on  @  \\\\\\"prop\\\\\\"                                        \\\\n  Retrieved descriptor              @  { value: \\\\\\"prop_value\\\\\\", writable: true, ... }  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log getPrototypeOf 1`] = `"[\\"\\\\n  Logger                  @  TGetPrototypeOf  \\\\n  Call getPrototypeOf of  @  TGetPrototypeOf  \\\\n  Retrieved prototype     @  Object           \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log has 1`] = `"[\\"\\\\n  Logger              @  TLogFullProperty            \\\\n  Access in operator  @  \\\\\\"prop\\\\\\" in TLogFullProperty  \\\\n  Access result       @  true                        \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log isExtensible 1`] = `"[\\"\\\\n  Logger                 @  TIsExtensible  \\\\n  Check isExtensible of  @  TIsExtensible  \\\\n  Result                 @  true           \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log only call events on methods 1`] = `"[\\"\\\\n  Logger       @  T                  \\\\n  Call method  @  method(): String;  \\\\n  Call result  @  \\\\\\"method_value\\\\\\"     \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log only get/set events on getters and setters 1`] = `"[\\"\\\\n  Logger           @  T                          \\\\n  Access property  @  get getSetProp(): String;  \\\\n  Access result    @  \\\\\\"getter_getSetProp\\\\\\"        \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log only get/set events on getters and setters 2`] = `"[\\"\\\\n  Logger        @  T                             \\\\n  Set property  @  set getSetProp(arg: String);  \\\\n  Set succeed   @  getSetProp = \\\\\\"\\\\\\";              \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log only get/set events on property 1`] = `"[\\"\\\\n  Logger           @  T              \\\\n  Access property  @  prop: String;  \\\\n  Access result    @  \\\\\\"prop_value\\\\\\"   \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log only get/set events on property 2`] = `"[\\"\\\\n  Logger        @  T              \\\\n  Set property  @  prop: String;  \\\\n  Set succeed   @  prop = \\\\\\"\\\\\\";     \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log ownKeys 1`] = `"[\\"\\\\n  Logger           @  TOwnKeys                    \\\\n  Get own keys of  @  TOwnKeys                    \\\\n  Result           @  [\\\\\\"key1\\\\\\", \\\\\\"key2\\\\\\", Symbol()]  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log ownKeys 2`] = `"[\\"\\\\n  Logger           @  TOwnKeys                    \\\\n  Get own keys of  @  TOwnKeys                    \\\\n  Result           @  [\\\\\\"key1\\\\\\", \\\\\\"key2\\\\\\", Symbol()]  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log ownKeys 3`] = `"[\\"\\\\n  Logger           @  TOwnKeys                    \\\\n  Get own keys of  @  TOwnKeys                    \\\\n  Result           @  [\\\\\\"key1\\\\\\", \\\\\\"key2\\\\\\", Symbol()]  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log ownKeys 4`] = `"[\\"\\\\n  Logger           @  TOwnKeys                    \\\\n  Get own keys of  @  TOwnKeys                    \\\\n  Result           @  [\\\\\\"key1\\\\\\", \\\\\\"key2\\\\\\", Symbol()]  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log ownKeys 5`] = `"[\\"\\\\n  Logger           @  TOwnKeys                    \\\\n  Get own keys of  @  TOwnKeys                    \\\\n  Result           @  [\\\\\\"key1\\\\\\", \\\\\\"key2\\\\\\", Symbol()]  \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log preventExtensions 1`] = `"[\\"\\\\n  Logger                     @  TPreventExtensions  \\\\n  Call preventExtensions on  @  TPreventExtensions  \\\\n  Result                     @  true                \\"]"`;

exports[`log/Log test when used as a class decorator should correctly log setPrototypeOf 1`] = `"[\\"\\\\n  Logger                  @  TSetPrototypeOf  \\\\n  Call setPrototypeOf on  @  TSetPrototypeOf  \\\\n  Set succeed             @  Object           \\"]"`;
